File: src-tauri/src/textutil.rs
//! macOS's own Word/RTF importer — the one behind TextEdit — at `/usr/bin/textutil`.
//!
//! WHY THIS EXISTS AS A SHARED MODULE: the legacy `.doc` path needs the same
//! converter twice, for two different outputs. The PREVIEW wants HTML (fonts,
//! weights, colours, lists); the EXTRACTED TEXT — what search, RAG and the
//! editor all read — wants plain text. Those used to come from two unrelated
//! places, and only the preview was any good: the text came from a `strings(1)`
//! style sweep of the OLE compound file that returned the font table and binary
//! noise. One document, two answers that disagreed.
//!
//! textutil ships with every Mac at a fixed path, works offline, and is nothing
//! to bundle, sign or notarize.

/// The formats `textutil` can import. `.docx` is deliberately absent from the
/// PREVIEW path (docx-preview renders more) but it can still be converted, so
/// this list is about what the importer understands, not about who uses it.
pub fn can_read(ext: &str) -> bool {
    matches!(ext, "doc" | "rtf" | "rtfd" | "odt" | "wordml" | "webarchive")
}

/// Convert `bytes` to `to` ("txt" or "html"). `None` when the format isn't
/// importable, the converter fails, or the result is empty.
///
/// Same bargain as the Quick Look bridge and the audio decoder: a system
/// converter takes a path, so the decrypted bytes touch the disk for the moment
/// it runs — owner-only, created exclusively — and are removed straight after.
pub fn convert(name: &str, bytes: &[u8], to: &str) -> Option<String> {
    let ext = crate::extraction::extension_of(name);
    if !can_read(&ext) {
        return None;
    }
    let stem = uuid::Uuid::new_v4();
    let dir = std::env::temp_dir();
    // Both paths are removed by Drop, so an early return, a panic in the
    // converter or a read failure can't leave decrypted bytes on disk.
    let src = TempPath(dir.join(format!("arcelle-tu-{stem}.{ext}")));
    let dst = TempPath(dir.join(format!("arcelle-tu-{stem}.{to}")));
    write_private(&src.0, bytes).ok()?;
    let run = std::process::Command::new("/usr/bin/textutil")
        .arg("-convert")
        .arg(to)
        .arg("-output")
        .arg(&dst.0)
        .arg(&src.0)
        .output();
    let out = match run {
        Ok(o) if o.status.success() => std::fs::read_to_string(&dst.0).ok(),
        _ => None,
    };
    out.filter(|h| !h.trim().is_empty())
}

/// A path that deletes itself. The decrypted copy this module writes must not
/// outlive the conversion, on any exit path.
struct TempPath(std::path::PathBuf);

impl Drop for TempPath {
    fn drop(&mut self) {
        let _ = std::fs::remove_file(&self.0);
    }
}

fn write_private(path: &std::path::Path, bytes: &[u8]) -> std::io::Result<()> {
    use std::io::Write;
    let mut opts = std::fs::OpenOptions::new();
    opts.write(true).create_new(true);
    #[cfg(unix)]
    {
        use std::os::unix::fs::OpenOptionsExt;
        opts.mode(0o600);
    }
    opts.open(path)?.write_all(bytes)
}

/// Turn Word FIELD CODES that survived the import into what they mean.
///
/// A `.doc` stores a hyperlink as a field: the instruction `HYPERLINK "url"`
/// followed by the display text. textutil imports the instruction as literal
/// characters, so the reader saw
/// `HYPERLINK "https://products.office.com/en-us/word"Mauris id ex erat.`
/// — the machinery of the document leaking into its prose, and the actual link
/// not a link at all.
///
/// Only the codes that carry user-visible meaning are handled. Everything else
/// is left exactly as it came, because guessing at an unknown field is how a
/// converter starts eating real text.
pub fn resolve_field_codes(text: &str, as_html: bool) -> String {
    // ` HYPERLINK "target" ` — optionally preceded by the field-start char and
    // followed by switches like \l "anchor" or \o "tooltip".
    let mut out = String::with_capacity(text.len());
    let mut rest = text;
    while let Some(at) = rest.find("HYPERLINK") {
        let (before, tail) = rest.split_at(at);
        let after = &tail["HYPERLINK".len()..];
        // The target is the first quoted string after the keyword.
        let Some(q1) = after.find('"') else {
            out.push_str(before);
            out.push_str("HYPERLINK");
            rest = after;
            continue;
        };
        let Some(q2rel) = after[q1 + 1..].find('"') else {
            out.push_str(before);
            out.push_str("HYPERLINK");
            rest = after;
            continue;
        };
        // Only treat it as a field if nothing but whitespace/switches sit
        // between the keyword and the quote — otherwise the word "HYPERLINK"
        // is just a word someone wrote.
        let gap = &after[..q1];
        if !gap.chars().all(|c| c.is_whitespace() || c == '\\' || c.is_alphanumeric()) {
            out.push_str(before);
            out.push_str("HYPERLINK");
            rest = after;
            continue;
        }
        let target = &after[q1 + 1..q1 + 1 + q2rel];
        out.push_str(before);
        // Trailing whitespace before the field belongs to the prose, not the link.
        if as_html && looks_like_url(target) {
            out.push_str(&format!(
                "<a href=\"{}\">{}</a>",
                escape_attr(target),
                escape_text(target)
            ));
        } else if !as_html {
            out.push_str(target);
        } else {
            out.push_str(&escape_text(target));
        }
        rest = &after[q1 + 1 + q2rel + 1..];
    }
    out.push_str(rest);
    out
}

/// Conservative: only schemes a document viewer should ever make clickable.
/// `javascript:` and friends must never become an `href`.
fn looks_like_url(s: &str) -> bool {
    let lower = s.trim().to_ascii_lowercase();
    lower.starts_with("https://") || lower.starts_with("http://") || lower.starts_with("mailto:")
}

fn escape_attr(s: &str) -> String {
    s.replace('&', "&amp;").replace('"', "&quot;").replace('<', "&lt;").replace('>', "&gt;")
}

fn escape_text(s: &str) -> String {
    s.replace('&', "&amp;").replace('<', "&lt;").replace('>', "&gt;")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn only_the_formats_macos_can_import_are_offered() {
        assert!(can_read("doc"));
        assert!(can_read("rtf"));
        // .docx is deliberately absent: it renders through docx-preview with
        // page breaks, headers, footers and images, which is more than this
        // importer gives.
        assert!(!can_read("docx"));
        assert!(!can_read("pptx"));
        assert!(!can_read("pdf"));
    }

    #[test]
    fn a_real_rtf_converts_to_text() {
        let rtf = br"{\rtf1\ansi{\fonttbl\f0\froman Times;}\f0\fs48 Hello \b bold\b0  world.\par}";
        let txt = convert("sample.rtf", rtf, "txt").expect("no text from textutil");
        assert!(txt.contains("Hello bold world."), "{txt}");
    }

    #[test]
    fn the_decrypted_temp_copy_deletes_itself_on_every_exit_path() {
        // Asserted on the guard rather than by counting `arcelle-tu-*` files in
        // the shared temp dir: the tests run in parallel, so that count is
        // whatever the OTHER tests happen to be doing at the time.
        let path = std::env::temp_dir().join(format!("arcelle-tu-{}.probe", uuid::Uuid::new_v4()));
        {
            let guard = TempPath(path.clone());
            write_private(&guard.0, b"decrypted bytes").expect("write");
            assert!(path.exists(), "the guard did not create its file");
        }
        assert!(!path.exists(), "a decrypted temp copy outlived the conversion");
    }

    #[test]
    fn the_temp_copy_is_owner_only() {
        // These bytes are the plaintext of an encrypted room; no other account
        // on the Mac may read them while the converter runs.
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let path =
                std::env::temp_dir().join(format!("arcelle-tu-{}.probe", uuid::Uuid::new_v4()));
            let guard = TempPath(path.clone());
            write_private(&guard.0, b"secret").expect("write");
            let mode = std::fs::metadata(&path).unwrap().permissions().mode();
            assert_eq!(mode & 0o077, 0, "temp copy readable by others: {mode:o}");
        }
    }

    #[test]
    fn a_real_rtf_converts_to_html_with_its_formatting() {
        let rtf = br"{\rtf1\ansi{\fonttbl\f0\froman Times;}\f0\fs48 Hello \b bold\b0  world.\par}";
        let html = convert("sample.rtf", rtf, "html").expect("no html");
        assert!(html.to_lowercase().contains("<b>"), "bold was lost: {html}");
    }

    #[test]
    fn a_hyperlink_field_becomes_the_link_and_stops_being_prose() {
        // The exact shape live QA found in a real .doc.
        let src = " HYPERLINK \"https://products.office.com/en-us/word\"Mauris id ex erat.";
        let text = resolve_field_codes(src, false);
        assert!(!text.contains("HYPERLINK"), "the field code leaked: {text}");
        assert!(text.contains("Mauris id ex erat."), "prose was eaten: {text}");
        assert!(text.contains("https://products.office.com"), "target lost: {text}");

        let html = resolve_field_codes(src, true);
        assert!(html.contains("<a href=\"https://products.office.com/en-us/word\">"), "{html}");
        assert!(html.contains("Mauris id ex erat."), "prose was eaten: {html}");
    }

    #[test]
    fn several_hyperlinks_in_one_document_all_resolve() {
        let src = "a HYPERLINK \"https://one.example\"one b HYPERLINK \"https://two.example\"two";
        let text = resolve_field_codes(src, false);
        assert!(!text.contains("HYPERLINK"), "{text}");
        assert!(text.contains("one.example") && text.contains("two.example"), "{text}");
        assert!(text.contains(" b "), "prose between the fields was lost: {text}");
    }

    #[test]
    fn a_non_http_target_never_becomes_a_clickable_href() {
        // A document must not be able to smuggle script into the reader.
        let html = resolve_field_codes("HYPERLINK \"javascript:alert(1)\"click", true);
        assert!(!html.contains("<a href"), "a script URL became a link: {html}");
        assert!(!html.contains("javascript:alert(1)\""), "unescaped: {html}");
        assert!(html.contains("click"), "prose was eaten: {html}");
    }

    #[test]
    fn the_word_hyperlink_in_ordinary_prose_is_left_alone() {
        let src = "The HYPERLINK, as it is called, points elsewhere.";
        assert_eq!(resolve_field_codes(src, false), src);
    }

    #[test]
    fn text_with_no_fields_is_returned_unchanged() {
        let src = "Ordinary prose with \"quotes\" and no fields at all.";
        assert_eq!(resolve_field_codes(src, false), src);
    }
}

File: apps/desktop/src/main/textUtil.ts
/**
 * Port of `src-tauri/src/textutil.rs` (296 lines, read in full) — the bridge
 * to macOS's own Word/RTF importer at the fixed path `/usr/bin/textutil` (the
 * one behind TextEdit and Quick Look), plus the Word FIELD CODE resolver that
 * turns a `HYPERLINK "url"displayed text` instruction that survived the
 * import into either a real `<a href>` or its plain-text target.
 *
 * WHY THIS EXISTS AS A SHARED MODULE (verbatim from the Rust source's own
 * header): the legacy `.doc` path needs the same converter twice, for two
 * different outputs. The PREVIEW wants HTML (fonts, weights, colours,
 * lists); the EXTRACTED TEXT — what search, RAG and the editor all read —
 * wants plain text. Those used to come from two unrelated places, and only
 * the preview was any good: the text came from a `strings(1)`-style sweep of
 * the OLE compound file that returned the font table and binary noise. One
 * document, two answers that disagreed. `textutil` ships with every Mac at a
 * fixed path, works offline, and is nothing to bundle, sign or notarize.
 *
 * ============================================================================
 * NOT A DUPLICATE OF `textClamp.ts` — checked before writing a line here.
 * ============================================================================
 * `textClamp.ts` ports a completely different Rust source
 * (`src-tauri/src/commands/agent.rs`'s `normalize_for_match`/`clamp_*`/
 * `excerpt` family): byte- and char-safe truncation of text a MODEL already
 * has in hand, for quote-matching and inventory descriptions. This file
 * ports `src-tauri/src/textutil.rs`: a macOS system-binary bridge (spawns a
 * real subprocess against a decrypted temp copy) plus a narrow HTML
 * field-code fixer. Neither the inputs, the outputs, nor a single function
 * name overlap between the two Rust files or these two ports — read both
 * files in full and confirmed there is no shared logic to fold together.
 *
 * ============================================================================
 * NOT A MODEL TOOL
 * ============================================================================
 * Grepped `toolSpecs.ts`/`toolSchema.ts`/`execTool.ts`: nothing there
 * references `textutil`, `office_html`, or `legacy` — this Rust module has no
 * `exec_tool` arm in `agent.rs` either (grepped the Rust tree the same way).
 * It is consumed internally by two Rust callers, `commands::office_html`
 * (`src-tauri/src/commands/office.rs`) and `extraction::legacy`
 * (`src-tauri/src/extraction/legacy.rs`) — a UI preview command and an
 * extraction-pipeline fallback, neither of which has an Electron port yet in
 * this tree (no `office.ts`/`legacy.ts` exists here). So this batch adds
 * nothing to `execTool.ts`, per this migration's rule 5 — there is no arm to
 * add additively because no arm calls this module yet.
 *
 * ============================================================================
 * A NOTE ON WHERE THIS CODE LIVES
 * ============================================================================
 * The migration design routes `textutil.rs` to the PYTHON SIDECAR under its
 * "document parsing goes to Python" split rule. That split does not
 * (yet) exist as running code anywhere in this tree — there is no Python
 * sidecar directory under `electron-migration/` at all, and several of its
 * own "sidecar" candidates (`editMatchDocx.ts`, `editMatchHtml.ts`,
 * `extractionWindow.ts`, `docsHtml.ts`, `previewTools.ts`'s non-PyObjC half)
 * already live in `src/main/*.ts` today. This port follows that same,
 * already-established precedent and lands in Electron main per this task's
 * explicit instruction. Unlike `quicklook.rs` (genuine PyObjC/Vision
 * dependency, stubbed NOT_IMPLEMENTED in `previewTools.ts`), `/usr/bin/
 * textutil` is an ordinary fixed-path CLI with no native-framework
 * dependency Node cannot reach directly — so this is a REAL, WORKING port,
 * not a stub. Flagging the plan/actual divergence for the owner rather than
 * silently resolving it either way.
 */

import { execFile } from "node:child_process";
import { randomUUID } from "node:crypto";
import * as fs from "node:fs";
import * as fsp from "node:fs/promises";
import * as os from "node:os";
import * as path from "node:path";

// ------------------------------------------------------------------ formats

/**
 * The formats `textutil` can import. `.docx` is deliberately absent from the
 * PREVIEW path (docx-preview renders more) but it can still be converted, so
 * this list is about what the importer understands, not about who uses it.
 * Ported verbatim from `textutil.rs`'s `can_read`.
 */
export function canRead(ext: string): boolean {
  return (
    ext === "doc" ||
    ext === "rtf" ||
    ext === "rtfd" ||
    ext === "odt" ||
    ext === "wordml" ||
    ext === "webarchive"
  );
}

/**
 * `extraction::extension_of` (`Path::new(name).extension()`, lower-cased),
 * including its own quirk: no dot, or a leading dot with no other dot (a
 * dotfile), both read as no extension at all. ANOTHER local copy — same as
 * `docsHtml.ts`'s and `organize.ts`'s own, for the same reason: this rewrite
 * has no shared `extraction.ts` yet for any of them to import instead.
 */
function extensionOf(name: string): string {
  const base = name.split(/[/\\]/).pop() ?? name;
  const idx = base.lastIndexOf(".");
  return idx <= 0 ? "" : base.slice(idx + 1).toLowerCase();
}

// -------------------------------------------------------------- temp hygiene

/**
 * `write_private`, ported verbatim: create `filePath` EXCLUSIVELY (fails if
 * it already exists — Rust's `create_new(true)`) and owner-only (`0o600`,
 * Rust's `.mode(0o600)`; both are still subject to the process umask on
 * either side, so both end up with the same effective bits), then write
 * `bytes`. Returns `false` on any failure rather than throwing, mirroring
 * `io::Result<()>` collapsed by the caller's `.ok()?`.
 *
 * Exported only so the delete-on-every-exit-path and owner-only-permission
 * guarantees can be tested directly against the mechanism, exactly as the
 * Rust source's own `#[cfg(test)] mod tests` reaches `write_private`/
 * `TempPath` in the same module rather than inferring the guarantee
 * indirectly through `convert()` — the Rust test's own comment explains why:
 * the shared temp dir is used by tests running in parallel, so counting
 * files there proves nothing about any one of them.
 */
export function writePrivate(filePath: string, bytes: Buffer): boolean {
  try {
    fs.writeFileSync(filePath, bytes, { flag: "wx", mode: 0o600 });
    return true;
  } catch {
    return false;
  }
}

/**
 * Delete `filePath`, discarding any error — Rust's `let _ =
 * std::fs::remove_file(&self.0)` inside `TempPath`'s `Drop`. A path that was
 * never created (e.g. because `writePrivate` failed before ever reaching the
 * OS write) is not an error here, matching `remove_file` on a missing file
 * being silently swallowed by the Rust guard. Exported for the same testing
 * reason as {@link writePrivate}.
 */
export function removeQuietly(filePath: string): void {
  try {
    fs.unlinkSync(filePath);
  } catch {
    // Best-effort by design — see doc comment.
  }
}

/**
 * `std::fs::read_to_string`, which is NOT what `readFile(path, "utf8")` does.
 * Rust REFUSES a file that is not valid UTF-8 (`io::ErrorKind::InvalidData`),
 * which the caller's `.ok()?` turns into "no text from this converter" and
 * the extraction pipeline treats as a reason to fall back. Node instead
 * substitutes U+FFFD for every bad byte and hands back a long, non-empty,
 * entirely meaningless string — which `convert`'s "not blank" filter passes
 * straight through to search, RAG and the editor as if it were the
 * document's prose. The second existing port of this same Rust source,
 * `services/agent-sidecar/src/arcelle_sidecar/docs/textutil.py`, is explicit about it too
 * (`except (OSError, UnicodeDecodeError): return None`).
 *
 * `ignoreBOM: true` means "do not strip a leading U+FEFF" — Rust keeps it as
 * a character, so this keeps it too.
 */
const STRICT_UTF8 = new TextDecoder("utf-8", { fatal: true, ignoreBOM: true });

/** Run `/usr/bin/textutil` with `args`, resolving `true` only on a clean
 * exit — a spawn failure, a non-zero exit, or a killing signal all resolve
 * `false`, mirroring the Rust source's `match run { Ok(o) if
 * o.status.success() => …, _ => None }`. */
function runTextutil(args: string[]): Promise<boolean> {
  return new Promise((resolve) => {
    execFile("/usr/bin/textutil", args, (error) => {
      resolve(error === null);
    });
  });
}

/**
 * Convert `bytes` (`name`'s decrypted content) to `to` (`"txt"` or `"html"`).
 * `null` when the format isn't importable, the converter fails, or the
 * result is empty after trimming — ported verbatim from `textutil.rs`'s
 * `convert`.
 *
 * Same bargain as the Quick Look bridge and the audio decoder (per the Rust
 * doc comment): a system converter takes a path, so the decrypted bytes
 * touch the disk for the moment it runs — owner-only, created exclusively —
 * and are removed straight after, on EVERY exit path (success, a failed
 * write, a failed convert, an empty result, or a thrown error), via the
 * `finally` below standing in for the Rust guard's `Drop`.
 */
export async function convert(name: string, bytes: Buffer, to: string): Promise<string | null> {
  const ext = extensionOf(name);
  if (!canRead(ext)) {
    return null;
  }
  const stem = randomUUID();
  const dir = os.tmpdir();
  const srcPath = path.join(dir, `arcelle-tu-${stem}.${ext}`);
  const dstPath = path.join(dir, `arcelle-tu-${stem}.${to}`);
  try {
    if (!writePrivate(srcPath, bytes)) {
      return null;
    }
    const ok = await runTextutil(["-convert", to, "-output", dstPath, srcPath]);
    if (!ok) {
      return null;
    }
    let out: string;
    try {
      // Read BYTES and decode strictly — see {@link STRICT_UTF8}. A decode
      // failure lands in the same `catch` as a missing/unreadable file, which
      // is exactly where Rust's `read_to_string(...).ok()` puts both.
      out = STRICT_UTF8.decode(await fsp.readFile(dstPath));
    } catch {
      return null;
    }
    return out.trim().length === 0 ? null : out;
  } finally {
    removeQuietly(srcPath);
    removeQuietly(dstPath);
  }
}

// --------------------------------------------------------- Word field codes

/** How much may sit between the `HYPERLINK` keyword and its target — room
 * for a space and a switch, and nothing like a clause of prose. Measured in
 * UTF-8 BYTES via `Buffer.byteLength`, matching the Rust source's
 * byte-counted `gap.len()` exactly (this is a length comparison, not a
 * content slice, so — unlike the ASCII-delimited slices around it — a
 * non-ASCII gap really could disagree between a byte count and a UTF-16
 * count). */
const MAX_FIELD_GAP = 8;

const HYPERLINK = "HYPERLINK";

/**
 * Rust's `str::split_whitespace` splits on `char::is_whitespace`, i.e. the
 * Unicode White_Space property — which is NOT the same set as JavaScript's
 * `\s`, and both differences flip this function's answer:
 *
 * - `\s` matches U+FEFF, which has NO White_Space property. Splitting on
 *   `/\s+/` made a stray BOM in the gap vanish, leaving zero tokens, so the
 *   vacuous `.every()` passed and `HYPERLINK﻿"great" and the rest.`
 *   came back as `great and the rest.` — the keyword and the quoted phrase
 *   deleted from the prose. That is precisely the failure the gap check
 *   exists to prevent, and a stray BOM is ordinary debris in text that has
 *   been through a legacy converter.
 * - `\s` does NOT match U+0085 (NEL), which DOES have the property. Rust
 *   sees no tokens and resolves the field; `/\s+/` saw a token that does not
 *   start with a backslash and left a genuine hyperlink unresolved.
 *
 * So the class is spelled out to be exactly Unicode White_Space.
 */
const RUST_WHITESPACE =
  /[\t\n\v\f\r \u0085\u00a0\u1680\u2000-\u200a\u2028\u2029\u202f\u205f\u3000]+/;

/**
 * Turn Word FIELD CODES that survived the import into what they mean.
 * Ported verbatim from `textutil.rs`'s `resolve_field_codes`, including its
 * exact bail conditions.
 *
 * A `.doc` stores a hyperlink as a field: the instruction `HYPERLINK "url"`
 * followed by the display text, with no separator between the closing quote
 * and that text. textutil imports the instruction as literal characters, so
 * the reader saw `HYPERLINK "https://products.office.com/en-us/word"Mauris
 * id ex erat.` — the machinery of the document leaking into its prose, and
 * the actual link not a link at all.
 *
 * Only the codes that carry user-visible meaning are handled — the gap
 * between the keyword and the quote must look like the rest of a field
 * instruction (short, and every whitespace-separated token in it a `\switch`
 * like `\l`/`\o`) or the word "HYPERLINK" is left exactly as it came, because
 * guessing at an unknown field is how a converter starts eating real text.
 *
 * BYTE-VS-CHAR NOTE: every slice point here (`"HYPERLINK"`, `"`) is an ASCII
 * delimiter, so slicing a JS string by UTF-16 index at that point yields the
 * identical Unicode content Rust's byte-index slicing would — the two only
 * disagree on how a LENGTH is counted, which is why {@link MAX_FIELD_GAP}'s
 * comparison alone uses `Buffer.byteLength`.
 */
export function resolveFieldCodes(text: string, asHtml: boolean): string {
  let out = "";
  let rest = text;
  for (;;) {
    const field = nextHyperlinkField(rest);
    if (field === null) break;
    out += field.before;
    if (field.target !== null) out += renderedFieldTarget(field.target, asHtml);
    rest = field.rest;
  }
  out += rest;
  return out;
}

interface HyperlinkField {
  readonly before: string;
  readonly target: string | null;
  readonly rest: string;
}

function nextHyperlinkField(text: string): HyperlinkField | null {
  const at = text.indexOf(HYPERLINK);
  if (at === -1) return null;
  const before = text.slice(0, at);
  const after = text.slice(at + HYPERLINK.length);
  const quotes = fieldQuotePositions(after);
  if (quotes === null || !isHyperlinkFieldGap(after.slice(0, quotes.first))) {
    return { before: before + HYPERLINK, target: null, rest: after };
  }
  const target = after.slice(quotes.first + 1, quotes.first + 1 + quotes.secondRelative);
  return { before, target, rest: after.slice(quotes.first + quotes.secondRelative + 2) };
}

interface FieldQuotePositions {
  readonly first: number;
  readonly secondRelative: number;
}

function fieldQuotePositions(text: string): FieldQuotePositions | null {
  const first = text.indexOf('"');
  if (first === -1) return null;
  const secondRelative = text.slice(first + 1).indexOf('"');
  return secondRelative === -1 ? null : { first, secondRelative };
}

function isHyperlinkFieldGap(gap: string): boolean {
  // Only treat it as a field if what sits between the keyword and the quote
  // looks like the rest of a field instruction — otherwise the word
  // "HYPERLINK" is just a word someone wrote. Everything Word puts there
  // before the target is a SWITCH, and every switch starts with a backslash.
  const tokens = gap.split(RUST_WHITESPACE).filter((token) => token.length > 0);
  return Buffer.byteLength(gap, "utf8") <= MAX_FIELD_GAP && tokens.every((token) => token.startsWith("\\"));
}

function renderedFieldTarget(target: string, asHtml: boolean): string {
  if (asHtml && looksLikeUrl(target)) return `<a href="${escapeAttr(target)}">${escapeText(target)}</a>`;
  return asHtml ? escapeText(target) : target;
}

/** Conservative: only schemes a document viewer should ever make clickable.
 * `javascript:` and friends must never become an `href`. Ported verbatim
 * from `looks_like_url`. Rust's `to_ascii_lowercase` lowercases ASCII bytes
 * only; JS's `toLowerCase()` also folds non-ASCII casing, but every prefix
 * checked below is pure ASCII, so the two never disagree on the boolean this
 * returns. */
function looksLikeUrl(s: string): boolean {
  const lower = s.trim().toLowerCase();
  return lower.startsWith("https://") || lower.startsWith("http://") || lower.startsWith("mailto:");
}

/** Ported verbatim from `escape_attr`. */
function escapeAttr(s: string): string {
  return s.replace(/&/g, "&amp;").replace(/"/g, "&quot;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

/** Ported verbatim from `escape_text`. */
function escapeText(s: string): string {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}
